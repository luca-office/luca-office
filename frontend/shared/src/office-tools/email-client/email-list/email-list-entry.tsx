import {css} from "@emotion/react"
import * as React from "react"
import {Icon, Tooltip} from "../../../components"
import {IconName} from "../../../enums"
import {
  backgroundColorBright,
  borderRadius,
  borderRadiusSmall,
  boxHeightMedium,
  deepShadow,
  Flex,
  fontColor,
  fontColorBright,
  fontColorLight,
  fontFamily,
  listEntrySelectedColor,
  primaryColor,
  spacing,
  spacingHuge,
  spacingSmall,
  spacingTinier,
  TextSize
} from "../../../styles"
import {useLucaTranslation} from "../../../translations"

export interface Props {
  isSelected: boolean
  isRead: boolean
  eMailAddress: string
  subject: string
  subLabel?: string
  onClick: () => void
  isIntroductionEmail?: boolean
}

export const EmailListEntry: React.FC<Props> = ({
  isSelected,
  eMailAddress,
  subject,
  subLabel,
  isRead,
  onClick,
  isIntroductionEmail = false
}) => {
  const {t} = useLucaTranslation()

  return (
    <div css={[Flex.row, styles.container(isSelected)]} onClick={onClick}>
      {isIntroductionEmail && (
        <Tooltip customStyles={styles.emailIconContainer} title={t("email__introduction_email_label")}>
          <Icon name={IconName.Email} color={fontColorBright} />
        </Tooltip>
      )}
      <div css={[Flex.column, styles.contentContainer]}>
        <div css={styles.emailRow}>
          <span css={styles.emailLabel}>{eMailAddress}</span>
          <div css={styles.emailStatusWrapper}>
            <span css={styles.emailLabel}>{subLabel}</span>
            <div css={styles.emailStatus(isRead)} />
          </div>
        </div>
        <div css={styles.subject}>{subject}</div>
      </div>
    </div>
  )
}

const styles = {
  container: (isSelected: boolean) =>
    css({
      paddingLeft: boxHeightMedium,
      position: "relative",
      border: `2px solid ${isSelected ? primaryColor : "transparent"}`,
      borderRadius: borderRadius,
      cursor: "pointer",
      boxSizing: "border-box",
      backgroundColor: isSelected ? listEntrySelectedColor : backgroundColorBright,
      transition: "all 0.2s",
      marginBottom: spacingTinier,

      ":hover": {
        boxShadow: deepShadow
      }
    }),
  contentContainer: css({
    margin: spacingSmall,
    width: "100%"
  }),
  emailLabel: css({
    fontFamily: fontFamily,
    fontSize: TextSize.Medium,
    letterSpacing: "0.15px",
    lineHeight: spacing(20),
    color: fontColor
  }),
  subject: css({
    fontFamily: fontFamily,
    fontSize: TextSize.Small,
    letterSpacing: "0.13px",
    lineHeight: spacing(18),
    color: fontColorLight,
    whiteSpace: "nowrap",
    overflow: "hidden",
    textOverflow: "ellipsis"
  }),
  emailIconContainer: css(Flex.column, {
    position: "absolute",
    top: 0,
    left: 0,
    bottom: 0,
    justifyContent: "center",
    alignItems: "center",
    height: "100%",
    width: spacingHuge,
    backgroundColor: primaryColor,
    borderTopLeftRadius: borderRadiusSmall,
    borderBottomLeftRadius: borderRadiusSmall
  }),
  emailRow: css({
    display: "flex",
    flexDirection: "row",
    justifyContent: "space-between"
  }),
  emailStatusWrapper: css({
    display: "flex",
    justifyContent: "space-between",
    alignItems: "center"
  }),
  emailStatus: (isRead: boolean) =>
    css({
      marginLeft: spacingSmall,
      borderRadius: "100%",
      width: 8,
      height: 8,
      background: !isRead ? primaryColor : undefined
    })
}
