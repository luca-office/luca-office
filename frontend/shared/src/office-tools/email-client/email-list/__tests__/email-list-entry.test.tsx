import {shallow} from "enzyme"
import * as React from "react"
import {create} from "react-test-renderer"
import {EmailListEntry} from "../email-list-entry"

const requiredProps = {
  isSelected: false,
  isRead: false,
  eMailAddress: "dummy.mail@mail.de",
  subject: "dummy subject",
  subLabel: "10:21",
  onClick: jest.fn()
}

describe("email-list-entry", () => {
  it("renders correctly with required props", () => {
    const listEntry = <EmailListEntry {...requiredProps} />
    const tree = create(listEntry).toJSON()
    expect(tree).toMatchSnapshot()
  })

  it("renders correctly with required props (selected)", () => {
    const listEntry = <EmailListEntry {...requiredProps} isSelected={true} />
    const tree = create(listEntry).toJSON()
    expect(tree).toMatchSnapshot()
  })

  it("calls onClick function correctly", () => {
    const onClick = jest.fn()

    const listEntry = <EmailListEntry {...requiredProps} onClick={onClick} />
    const tree = shallow(listEntry)
    tree.simulate("click")
    expect(onClick).toBeCalledTimes(1)
  })
})
