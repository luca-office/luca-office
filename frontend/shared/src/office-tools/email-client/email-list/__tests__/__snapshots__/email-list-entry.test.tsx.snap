// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`email-list-entry renders correctly with required props (selected) 1`] = `
<div
  className="css-t3bqwd-EmailListEntry"
  onClick={[MockFunction]}
>
  <div
    className="css-5pl3qu-EmailListEntry"
  >
    <div
      className="css-t5swy5-EmailListEntry"
    >
      <span
        className="css-1qipgha-EmailListEntry"
      >
        dummy.mail@mail.de
      </span>
      <div
        className="css-klkx3-EmailListEntry"
      >
        <span
          className="css-1qipgha-EmailListEntry"
        >
          10:21
        </span>
        <div
          className="css-1ye5fak-EmailListEntry"
        />
      </div>
    </div>
    <div
      className="css-n6fk5y-EmailListEntry"
    >
      dummy subject
    </div>
  </div>
</div>
`;

exports[`email-list-entry renders correctly with required props 1`] = `
<div
  className="css-rc8o96-EmailListEntry"
  onClick={[MockFunction]}
>
  <div
    className="css-5pl3qu-EmailListEntry"
  >
    <div
      className="css-t5swy5-EmailListEntry"
    >
      <span
        className="css-1qipgha-EmailListEntry"
      >
        dummy.mail@mail.de
      </span>
      <div
        className="css-klkx3-EmailListEntry"
      >
        <span
          className="css-1qipgha-EmailListEntry"
        >
          10:21
        </span>
        <div
          className="css-1ye5fak-EmailListEntry"
        />
      </div>
    </div>
    <div
      className="css-n6fk5y-EmailListEntry"
    >
      dummy subject
    </div>
  </div>
</div>
`;
